<a href="https://colab.research.google.com/github/eray1261/Fake-News-Detection-System/blob/main/FakeNesDetectionSystem.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [26]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving news.csv to news.csv


In [ ]:
#Read the data
df=pd.read_csv('news.csv')

#Get shape and head
df.shape
df.head()

Unnamed: 0  ... label
0        8476  ...  FAKE
1       10294  ...  FAKE
2        3608  ...  REAL
3       10142  ...  FAKE
4         875  ...  REAL

[5 rows x 4 columns]

In [ ]:
labels=df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [23]:
df.dtypes

Unnamed: 0     int64
title         object
text          object
label         object
dtype: object

In [ ]:
X_train,X_test,Y_train,Y_test=train_test_split(df['text'], labels, test_size=0.2, random_state=7)

**TfidVectorizer:**

***TF (Term Frequency):*** It is the number of times a word appears in a document is its Term Frequency. A higher value means a term appears more often than others, and so, the document is a good match when the term is part of the search terms.

***IDF (Inverse Document Frequency):*** Words that occur many times a document, but also occur many times in many others, may be irrelevant. IDF is a measure of how significant a term is in the entire corpus.

The TfidfVectorizer converts a collection of raw documents into *a matrix of TF-IDF features*.

In [ ]:
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)
tfidf_train=tfidf_vectorizer.fit_transform(X_train) 
tfidf_test=tfidf_vectorizer.transform(X_test)

**PassiveAggressiveClassifier:**

Passive Aggressive algorithms are *online learning algorithms*. Such an algorithm remains passive for a correct classification outcome, and turns aggressive in the event of a miscalculation, updating and adjusting. Unlike most other algorithms, *it does not converge*. Its purpose is to make updates that correct the loss, causing very little change in the norm of the weight vector.



In [ ]:
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,Y_train)

y_pred=pac.predict(tfidf_test)
score=accuracy_score(Y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.74%


We got an accuracy of 92.82% with this model. Next, let’s print out a confusion matrix to gain insight into the number of false and true negatives and positives.

In [27]:
confusion_matrix(Y_test,y_pred, labels=['FAKE','REAL'])

array([[588,  50],
       [ 42, 587]])

So with this model, we have 588 true positives, 587 true negatives, 42 false positives, and 50 false negatives.